polyply/src/topology.py

```diff
@@ -356,41 +357,75 @@ def convert_nonbond_to_sig_eps(self):
 
             self.nonbond_params.update({atom_pair: {"nb1": sig, "nb2": eps}})
 
-    def add_positions_from_file(self, path, build_res=[]):
+    def add_positions_from_file(self, path, skip_res=[], resolution='mol'):
         """
-        Add positions to topology from coordinate file.
+        Add positions to molecules in topology from coordinate file.
+        Depending on the resolution set they are either added to the
+        meta_molecule or the molecules. With `skip_res` residues can
+        be skipped by name.
+
+        Note that molecule coordinates will also set the meta_molecule
+        coordinates. Coordinates for a single residue must be complete.
+
+        Currently .gro and .pdb file parsers are supoprted for the
+        coordinate reading. See `_coord_parsers` for more information.
+
+        Parameters
+        ----------
+        path: :class:`pathlib.Path`
+            path to coordinate file
+        skip_res: list[str]
+            list of resnames to skip
+        resolution: str
+            choice of meta_mol or mol
         """
         path = Path(path)
         extension = path.suffix.casefold()[1:]
-        molecules = _coord_parser(path, extension)
+        positions = _coord_parser(path, extension)
+        max_coords = len(positions)
         total = 0
         for meta_mol in self.molecules:
-            no_coords = []
-            for node in meta_mol.molecule.nodes:
-                resname = meta_mol.molecule.nodes[node]["resname"]
-                if resname in build_res:
-                    no_coords.append(node)
-                    continue
-                try:
-                    position = molecules.nodes[total]["position"]
-                except KeyError:
-                    no_coords.append(node)
-                else:
-                    meta_mol.molecule.nodes[node]["position"] = position
+            for meta_node in meta_mol.nodes:
+                resname = meta_mol.nodes[meta_node]["resname"]
+                mol_nodes = meta_mol.nodes[meta_node]['graph'].nodes
+                # skip residue if resname is to be skipped or
+                # if the no more coordinates are available
+                # in that case we want to build the node and
+                # backmap it
+                if resname in skip_res or total >= max_coords:
+                    meta_mol.nodes[meta_node]["build"] = True
+                    meta_mol.nodes[meta_node]["backmap"] = True
+                # here we only add meta_molecule coordiantes
+                # in that case we only want to backmap
+                elif resolution == 'meta_mol':
```

Comment:
Giving not enough positions for all the `meta_mol.nodes` could happen unintentionally by the user.
Maybe it is worth throwing a warning when `total >= max_coords`, since users intentionally doing so seems less likely to me.

Reply:
the error is not raised from the topology because this is a very polyply specific thing. in principle I like to keep the topology more clean in terms of API and logging. Thus the error is raised at the gen_coords level.